### PR TITLE
Request controllers

### PR DESCRIPTION
**Pull request by gagandeep95**
### What this PR does (required):
- This is the Request Controller which will get the requests for the logged-in user, create a new request and update the request with approved or declined status

### Any information needed to test this feature (required):
- In Postman (raw data (JSON)),  add the id of any registered user as sitterId, start date, end date to create the request.

### CODE REVIEW

File: server/controllers/request.js
Comment (ezeilo-su) [inline server/controllers/request.js:6]: Good documentation

File: server/controllers/request.js
Comment (ezeilo-su) [inline server/controllers/request.js:14]: :+1:

File: server/controllers/request.js
Comment (ezeilo-su) [inline server/controllers/request.js:17]: I don't see this as a good practice to modify request body, though. It's possible for other middleware to also have access to the `req` object, thereby having that information exposed to them.

File: server/controllers/request.js
Comment (ezeilo-su) [inline server/controllers/request.js:23]: :+1:

File: server/controllers/request.js
Comment (ezeilo-su) [inline server/controllers/request.js:32]: Make proper use of  `async/await`. There's no point making use of callback in an `async` function.

File: server/models/Request.js
Comment (ezeilo-su) [inline server/models/Request.js:34]: :+1:

File: server/routes/request.js
Comment (ezeilo-su) [inline server/routes/request.js:13]: :+1:

**Review by ezeilo-su — COMMENTED** [review]
Look nice :+1:  Just a few minor comments to take a look at.

File: server/controllers/request.js
Comment (gagandeep95) [reply server/controllers/request.js:17]: Thanks, @ezeilo-su , I will remove it from request body

File: server/routes/request.js
Comment (ugur-akin) [inline server/routes/request.js:13]: `PUT` verb implies a complete replacement of a resource, but your update controller uses `findByIdAndUpdate` method (which by default doesn't necessarily overwrite the entire resource, see the `overwrite` option [here](https://restfulapi.net/resource-naming/)). 

If, by design, you want to encourage the callers do partial updates, we can keep the controller as is and replace the verb with `PATCH`. Otherwise, if you expect a complete replacement of a resource sent with each request, then prefer the `findByIdAndReplace` method.

Finally, [this stackoverflow answer](https://stackoverflow.com/a/19736570) gives good insight on why the distinction might be important for us 👍🏼

File: server/models/Request.js
Comment (ugur-akin) [inline server/models/Request.js:13]: This file is not relevant to this PR. There are a couple of ways we can go about using Git when our features depend on other features in development:

- **For simpler cases**: If the depended updates are small enough (e.g. few files/folders) and aren't going to be mutually modified, we can manually check them out but not stage them (e.g. use `git checkout depended-feature-branch -- path(s)/to/files/folders` but do not `git add` them). These will create local copies of these files and we can develop/test our new feature without a problem.

- In most cases, the solution above won't be feasible (e.g. both branches update the same file(s)). A more robust approach is to branch off your new feature of the depended feature branch (e.g. `git checkout depended-feature && git branch new-feature`). The changes from the dependent feature are going to be present and tracked automatically in this case. 

For the latter approach, use `git rebase depended-branch` often during the development of the new feature to ensure your commit history is always up-to-date and you have all the latest changes from the depended branch. When you create a PR for the new feature, if the depended feature isn't merged in yet, it's advisable to change your base to `depended-branch` instead of `main` to facilitate the review process. Once the depended feature is merged in, we can use `git rebase main` to clean up our commit history and correctly reflect changes exclusive to the new feature (you might have to do a `git push --force-with-lease` in this case). We should then change the base branch back to main for the PR.

**Review by ugur-akin — CHANGES_REQUESTED** [review]
Looks solid overall 🙌🏼 I left some informational tips/suggestions here and there, feel free to further any discussions! 😇

File: server/controllers/request.js
Comment (gagandeep95) [reply server/controllers/request.js:26]: Hey @ugur-akin , Thanks for sharing these valuable suggestions. I made these changes.

File: server/models/Request.js
Comment (gagandeep95) [reply server/models/Request.js:13]: Thanks @ugur-akin , This is very helpful. I was not aware of branching off the depended feature. I will definitely use this strategy in the future.

**Comment by gagandeep95** [comment]
Hi @ugur-akin , 
For now, As my dependent feature branch (**RequestModel**) has been merged to main, and I did something like the below:
- Delete the  Model/Request.js file from Request Controllers (wanted to get Request Model code received from the main branch
- git checkout main
- git pull origin main
- git checkout RequestControllers
- git rebase main
 I did not receive the Request Model file in my RequestControllers.

**Comment by ugur-akin** [comment]
> Delete the Model/Request.js file from Request Controllers (wanted to get Request Model code received from the main branch
git checkout main
git pull origin main
git checkout RequestControllers
git rebase main
I did not receive the Request Model file in my RequestControllers. Why is this so?

Hi @gagandeep95 , if the commits on your `RequestControllers` branch include the deletion of the file (which seems to be the case), then rebase won't bring that file back as your commit that deleted the file will just be re-applied down the road. In this case, the simplest solution would be to recover the file from main and add a new commit with the file restored.

The above approach unfortunately leads to a dirtier commit history. A more advanced approach would be to go back into the commit where we deleted the file, re-apply the commits but exclude the deletion but this seems very complicated with merge commits along the way. I'll do more research on this to see if it is indeed feasible given the complex history of the branch 👍🏼

File: server/controllers/request.js
Comment (ugur-akin) [inline server/controllers/request.js:17]: This has eluded me before, but if we accept any user id here, any logged in user can create artificial requests for other owners - we should read the `req.user.id` field for the appropriate user id instead👍🏼

File: server/controllers/request.js
Comment (ugur-akin) [inline server/controllers/request.js:27]: For the same reason as above, we should try to prevent updates to user id (and potentially sitter id) field. We could validate the body content but a simpler approach would be to destructure the body to limit what fields this request can update.

File: server/controllers/request.js
Comment (ugur-akin) [inline server/controllers/request.js:22]: This is now used in a `PATCH` route 👍🏼

File: server/controllers/request.js
Comment (ugur-akin) [inline server/controllers/request.js:29]: 👍🏼

File: server/routes/request.js
Comment (ugur-akin) [inline server/routes/request.js:13]: 🙌🏼

File: server/controllers/request.js
Comment (ugur-akin) [inline server/controllers/request.js:26]: Alternatively, you can opt for the `findByIdAndUpdate` call here 👍🏼

**Review by ugur-akin — CHANGES_REQUESTED** [review]
Looks good, thanks for the changes. I noticed another potential security concern that would be ideal to fix 👍🏼

Regarding the deletion of `server/models/Request.js`, it's important to understand that deleting a file won't delete the history of that file on git - instead it will add one more change to it! So in order to undo unwanted/accidental changes, we will often have to resort to git commands such as reset, rm, revert, restore...

**Comment by gagandeep95** [comment]
> Looks good, thanks for the changes. I noticed another potential security concern that would be ideal to fix 👍🏼
> 
> Regarding the deletion of `server/models/Request.js`, it's important to understand that deleting a file won't delete the history of that file on git - instead it will add one more changes to it! So in order to undo unwanted/accidental changes, we will often have to resort to git commands such as reset, rm, revert, restore...


Hey @ugur-akin , Thanks for the suggestions. I tried to use this:  **git revert --no-commit dad279d** and it successfully retrieved the deleted file in the RequestControllers branch. However, when I again tried to do  **git rebase main**, It removed the Request Model File in my RequestControllers branch. I checked my main branch, this file is present there. Why it is deleting the file?

File: server/controllers/request.js
Comment (gagandeep95) [reply server/controllers/request.js:17]: Hi @ugur-akin,  Can you please explain this more?

Comment (gagandeep95) [reply server/controllers/request.js:17]: Hi @ugur-akin , Are you suggesting modifying req body in the controllers like below: 

```
const createRequest = async (req, res) => {
  req.body.userId = req.user.id;
const request = await Request.create(req.body);
  res.send(request);
}; 
```

File: server/controllers/request.js
Comment (gagandeep95) [reply server/controllers/request.js:27]: Thanks, I will make it done

**Comment by ugur-akin** [comment]
`--no-commit` will stage the changes necessary in your working directory - have you commited them manually? Git won't be able to apply those changes upon a rebase unless you make a commit. That's the only thing I can think of without investigating further.

File: server/controllers/request.js
Comment (ugur-akin) [reply server/controllers/request.js]: Not necessarily, modifying the body isn't necessary and could be misleading in some cases. For a quick fix you can do the following:
```
const request = await Request.create({...req.body, userId: req.user.id});
```

But ideally, we should instead destructure the body and explicitly pass the fields, e.g:
```
const { sitterId, start, end, ... } = req.body;
const userId = req.user.id;

const request = await Request.create({userId, sitterId, start, end, ...});
```

Comment (gagandeep95) [reply server/controllers/request.js]: Thanks @ugur-akin

File: server/controllers/request.js
Comment (ezeilo-su) [inline server/controllers/request.js:17]: A good security practice on this one :+1: It will also be good to validate request body so you can return a `400` on a missing field.

Comment (ezeilo-su) [inline server/controllers/request.js:17]: A good security practice on this one :+1: It will also be good to validate the request body so you can return a `400` on a missing field.

File: server/controllers/request.js
Comment (ezeilo-su) [inline server/controllers/request.js:27]: Also validate this one

File: server/models/Request.js
Comment (ezeilo-su) [inline server/models/Request.js:25]: :+1:

File: server/routes/request.js
Comment (ezeilo-su) [inline server/routes/request.js:13]: :+1:

**Review by ezeilo-su — COMMENTED** [review]
Good work with those changes. A few comments to take a look, though.

File: server/controllers/request.js
Comment (gagandeep95) [reply server/controllers/request.js:17]: Hi @ezeilo-su , I will use front-end validations for missing fields.

File: server/controllers/request.js
Comment (gagandeep95) [reply server/controllers/request.js:27]: @ezeilo-su , I think, this is from the previous commit. I already made the changes in this file.

File: server/models/Request.js
Comment (rajivtitus) [inline server/models/Request.js:24]: 👍🏽

File: server/controllers/request.js
Comment (rajivtitus) [inline server/controllers/request.js:9]:
```suggestion
  const requests = await Request.find({ $or: [{ ownerId: req.user.id }, { sitterId: req.user.id }] })

```
This query will only search for service requests which have been requested by the user, instead you can change the query as above :)

File: server/controllers/request.js
Comment (rajivtitus) [reply server/controllers/request.js:17]: @gagandeep95 It is a great practice to validate input parameters on the BE as well 😄

File: server/controllers/request.js
Comment (rajivtitus) [reply server/controllers/request.js:27]: Same as my comment above about validtion

**Review by rajivtitus — COMMENTED** [review]
Hey @gagandeep95, just a couple of small suggestions to address and this is good to go 👍🏽 
I am going to mark the validation of the input parameters as optional (would be nice to have though)

**Comment by rajivtitus** [comment]
Hey @gagandeep95, please address the comments above when you get a chance :)

File: server/controllers/request.js
Comment (rajivtitus) [inline server/controllers/request.js:32]:
```suggestion
    { _id: req.params.id, sitterId: userId },
```
This should be sitterId, as only sitters can accept/decline requests

**Review by rajivtitus — APPROVED** [review]
LGTM! Just address pending comment before merging :)

File: server/controllers/request.js
Comment (gagandeep95) [reply server/controllers/request.js:32]: @rajivtitus  Thanks

**Pull request merged**
